channels/rdpgfx/server/rdpgfx_main.c

```diff
@@ -0,0 +1,1239 @@
+	if ((error = rdpgfx_write_header(s, &header)))
+	{
+		WLog_ERR(TAG, "rdpgfx_write_header failed with error %lu!", error);
+		return NULL;
+	}
+
+	return s;
+}
+
+/**
+ * Function description
+ * Send the stream for rdpgfx packet. The packet would be compressed
+ * according to [MS-RDPEGFX].
+ *
+ * @return 0 on success, otherwise a Win32 error code
+ */
+static UINT rdpgfx_server_packet_send(RdpgfxServerContext* context, wStream* s)
+{
+	UINT error;
+	UINT32 flags = 0;
+	ULONG written;
+	BYTE* pSrcData = Stream_Buffer(s);
+	UINT32 SrcSize = Stream_GetPosition(s);
+	wStream* fs;
+	
+	/* Fill actual length */
+	Stream_SetPosition(s, sizeof(RDPGFX_HEADER) - sizeof(UINT32));
+	Stream_Write_UINT32(s, SrcSize); /* pduLength (4 bytes) */
+
+	/* Allocate new stream with enough capacity. Additional overhead is
+	 * descriptor (1 bytes) + segmentCount (2 bytes) + uncompressedSize (4 bytes)
+	 * + segmentCount * size (4 bytes) */
+	fs = Stream_New(NULL, SrcSize + 7 
+	   + (SrcSize/ZGFX_SEGMENTED_MAXSIZE + 1) * 4);
+
+	if (zgfx_compress_to_stream(context->priv->zgfx, fs, pSrcData, SrcSize, &flags) < 0)
+	{
+		WLog_ERR(TAG, "zgfx_compress_to_stream failed!");
+		error = ERROR_INTERNAL_ERROR;
+		goto out;
+	}
+
+	error = WTSVirtualChannelWrite(context->priv->rdpgfx_channel, (PCHAR) Stream_Buffer(fs), 
+								   Stream_GetPosition(fs), &written) 
+								   ? CHANNEL_RC_OK : ERROR_INTERNAL_ERROR;
+
+out:
+	Stream_Free(fs, TRUE);
+	Stream_Free(s, TRUE);
+
+	return error;
+}
+
+/**
+ * Function description
+ *
+ * @return 0 on success, otherwise a Win32 error code
+ */
+static UINT rdpgfx_send_caps_confirm_pdu(RdpgfxServerContext* context, RDPGFX_CAPS_CONFIRM_PDU* capsConfirm)
+{
+	RDPGFX_CAPSET* capsSet = capsConfirm->capsSet;
+	wStream* s = rdpgfx_server_packet_new(
+		     RDPGFX_CMDID_CAPSCONFIRM, 
+		     sizeof(RDPGFX_CAPSET) + sizeof(capsSet->flags));
+
+	if (!s)
+	{
+		WLog_ERR(TAG, "rdpgfx_server_packet_new failed!");
+		return CHANNEL_RC_NO_MEMORY;
+	}
+
+	Stream_Write_UINT32(s, capsSet->version); /* version (4 bytes) */
+	Stream_Write_UINT32(s, sizeof(capsSet->flags)); /* capsDataLength (4 bytes) */
+	Stream_Write_UINT32(s, capsSet->flags); /* capsData (4 bytes) */
+
+	return rdpgfx_server_packet_send(context, s);
+}
+
+/**
+ * Function description
+ *
+ * @return 0 on success, otherwise a Win32 error code
+ */
+static UINT rdpgfx_send_reset_graphics_pdu(RdpgfxServerContext* context, RDPGFX_RESET_GRAPHICS_PDU* pdu)
+{
+	UINT32 index;
+	MONITOR_DEF* monitor;
+	wStream* s = rdpgfx_server_packet_new(
+		     RDPGFX_CMDID_RESETGRAPHICS,
+		     RDPGFX_RESET_GRAPHICS_PDU_SIZE - RDPGFX_HEADER_SIZE);
+
+	if (!s)
+	{
+		WLog_ERR(TAG, "rdpgfx_server_packet_new failed!");
+		return CHANNEL_RC_NO_MEMORY;
+	}
+
+	Stream_Write_UINT32(s, pdu->width); /* width (4 bytes) */
+	Stream_Write_UINT32(s, pdu->height); /* height (4 bytes) */
+	Stream_Write_UINT32(s, pdu->monitorCount); /* monitorCount (4 bytes) */
+
+	for (index = 0; index < pdu->monitorCount; index++)
```

Comment:
How can you be sure that s will be large enough to hold all monitors informations (if you have many) ?